### PR TITLE
Major Code Rewrite

### PR DESCRIPTION
**Pull request by OakLoaf**
## Changes
- Rewrote HologramData structure
- Added single schedule thread for holograms
  - This has also been implemented throughout a handful of API methods to ensure correct thread usage
- Moved EnableChecker class into FancyHologramsPlugin
- Cleaned up file tree
- Added more detailed error message when shading API instead of compiling only

_I am open to suggestions, and can understand if reverting the api file tree is requested_
_I have tested this on a local server and have not found any issues, this will of course need testing by others_

### CODE REVIEW

**Comment by bridgelol** [comment]
If you want to make this actually decent the whole system should be rewritten with a central "entity tracker", instead of submitting new tasks so often

**Comment by OakLoaf** [comment]
> If you want to make this actually decent the whole system should be rewritten with a central "entity tracker", instead of submitting new tasks so often

I do agree that this would be beneficial, I'd be down to do this

**Comment by OakLoaf** [comment]
I have not yet made changes to creating, removing, hiding or showing holograms to players. 

Regarding creating and removing I need to get my head around the current implementation before making any changes.

I'll likely make `showHologram` and `hideHologram` simply add/remove the player from the viewers list and then let the task in the HologramManager to do the showing and hiding.

File: api/src/main/java/de/oliver/fancyholograms/api/hologram/Hologram.java [hunk 1]
```diff
@@ -162,7 +149,7 @@ protected boolean shouldHologramBeShown(@NotNull final Player player) {
      * @param player the player to check and update the shown state for
      */
     public void checkAndUpdateShownStateForPlayer(Player player) {
```
[the remark this answers is not in the document]
Comment (OakLoaf) [reply api/src/main/java/de/oliver/fancyholograms/api/hologram/Hologram.java]: I agree, the initial run through of implementing the thread was assuming that the current implementation didn't have these sort of issues. I have now removed that

**Comment by OakLoaf** [comment]
I think this pr is relatively read to go now, the only thing I haven't done is changed creating and removing holograms but that could be done in a separate pr (Unless anyone has any immediate improvements that could be made)

**Comment by R00tB33rMan** [comment]
Can vouch that this performs exceptionally better than FancyHolograms upstream!

**Review by OliverSchlueter — APPROVED** [review]
LGTM

**Comment by OakLoaf** [comment]
This should now have no conflicts, the recent features `Visibility`, `TextColor -> Color` and `Persistency`  may need testing though.

**Pull request merged**